### ISSUE TITLE
Add Check Support for Java 21 Record Pattern : IllegalType

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14946]
child of #14890 
I have read check documentation: https://checkstyle.org/checks/coding/illegaltype.html#IllegalType
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/RecordPatterns.java
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

public class RecordPatterns {
    record Box<T>(T t) { }

    public void TestIllegalType() {
        List<LinkedHashMap<Integer, Integer>> l // violation
                = new ArrayList<>();

        if (l instanceof List<LinkedHashMap<Integer, Integer>> linkedHashMap) { // violation
            System.out.println(linkedHashMap);
        }
         LinkedHashMap<Integer, Integer> l2 // violation
                = new LinkedHashMap<>();
        Box<LinkedHashMap<Integer,Integer>> box = new Box<>(l2);  // violation

        if (box instanceof Box<LinkedHashMap<Integer,Integer>>(var linkedHashMap)) { // expected violation
            System.out.println(linkedHashMap);
        }
    }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="IllegalType"/>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:9:14: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:12:31: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:15:10: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:17:13: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 4 errors.
```
---

**Describe what you expect in detail.**

I expect a violation on the type parameters of the record pattern.  The type parameter of the record pattern is of illegal type. 

---

**Comment by rnveach** [comment]
I am fine with this issue.

> --enable-preview 

If this isn't a preview feature, so this shouldn't be needed right? I would remove this to avoid confusion.

> I expect a violation on the record pattern as we have a violation on the normal type pattern (line 12).
> `if (l instanceof List<LinkedHashMap<Integer, Integer>> linkedHashMap) { // violation`

The issue is slightly confusing as the line says missing a violation has a comment which says violation. I had to make sure I was finding which one was line 12.

My recommendation is make the line with a missing violation something else like "expected violation" or "no violation".

> `if (box instanceof Box<LinkedHashMap<Integer,Integer>>(var linkedHashMap)) {`

Your example only used `var` which is not a strict type, but we will be expanding this check for things like (random example from JEP):
`    if (r instanceof Rectangle(ColoredPoint(Point p, Color c),
                               ColoredPoint lr)) {`

for example if I wanted to forbid `Rectangle`, `ColoredPoint`, or even `Point`/`Color`?

**Comment by mahfouz72** [comment]
> The is slightly confusing as the line says missing a violation but has a comment which says violation. I had to make sure I was finding the right line 12.

Yes sorry my wording could be confusing. I wanted to say that I want a violation on record pattern (line 19)  like we have a violation on the type pattern (line 12). 

> Your example only used var which is not a strict type, but we will be expanding this check for things like (random example from JEP):

yes ofc this should be solved within this issue. I just was trying to make the example with the default config. here is extra example:

```Java
PS D:\CS\test> javac src/RecordPatterns.java                                               
PS D:\CS\test> cat src/RecordPatterns.java                                                 
record ColoredPoint(Point p, String c) { } // violation
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }  // 2 violations
record Point(int x, int y) {}

public class RecordPatterns {
    public void TestIllegalType(Object obj) {
        if (obj instanceof Rectangle r) { }  // violation
        if (obj instanceof ColoredPoint l)  // violation
        if (obj instanceof Rectangle(ColoredPoint x, ColoredPoint y)) { } // 2 violations, expected  3
        if (obj instanceof Rectangle(ColoredPoint(Point(int x1,int x2),_),_)) { } // no violations, expected 3
    }
}
PS D:\CS\test> cat config.xml                                                              
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="IllegalType">
        <property name="illegalClassNames" value="Rectangle, ColoredPoint, Point"/>
    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:1:21: Usage of type 'Point' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:2:18: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:2:42: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:7:28: Usage of type 'Rectangle' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:8:28: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:9:38: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:9:54: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 7 errors.

```

**Comment by nrmancuso** [comment]
I am good to approve after @rnveach is satisfied with items from https://github.com/checkstyle/checkstyle/issues/14946#issuecomment-2151326698

**Comment by nrmancuso** [comment]
@rnveach ping

**Comment by rnveach** [comment]
I am good, everything was satisified. Sorry for delay.
